compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala

```diff
@@ -1062,7 +1062,13 @@ class TreeUnpickler(reader: TastyReader, tastyName: TastyName.Table, posUnpickle
       }
 
     def readCases(end: Addr)(implicit ctx: Context): List[CaseDef] =
-      collectWhile(nextByte == CASEDEF && currentAddr != end) { readCase()(ctx.fresh.setNewScope) }
+      collectWhile((nextByte == CASEDEF || nextByte == SHARED) && currentAddr != end) {
```

Comment:
You can simply use `nextUnsharedTag` for this. 

    collectWhile(nextUnsharedTag == CASEDEF && currentAddr != end) { readCase()(ctx.fresh.setNewScope) }